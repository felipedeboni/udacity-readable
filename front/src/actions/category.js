export const FETCH_CATEGORIES = 'FETCH_CATEGORIES';
export const FETCH_CATEGORIES_SUCCESS = 'FETCH_CATEGORIES_SUCCESS';
export const FETCH_CATEGORIES_FAIL = 'FETCH_CATEGORIES_FAIL';

export const fetchCategories = () => {
  return {
    type: FETCH_CATEGORIES,
    payload: {
      request: {
        url: '/categories'
      }
    }
  };
};
