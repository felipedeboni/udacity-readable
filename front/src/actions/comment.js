import uuid from 'utils/uuid';

// GET /posts/:id/comments
// Get all the comments for a single post.
export const FETCH_COMMENTS_BY_POST = 'FETCH_COMMENTS_BY_POST';
export const FETCH_COMMENTS_BY_POST_SUCCESS = 'FETCH_COMMENTS_BY_POST_SUCCESS';
export const FETCH_COMMENTS_BY_POST_FAIL = 'FETCH_COMMENTS_BY_POST_FAIL';
export const fetchCommentsByPost = postId => {
  return {
    type: FETCH_COMMENTS_BY_POST,
    payload: {
      request: {
        url: `/posts/${postId}/comments`
      }
    }
  };
};

// GET /comments/:id
// Get the details for a single comment.
export const FETCH_COMMENT = 'FETCH_COMMENT';
export const FETCH_COMMENT_SUCCESS = 'FETCH_COMMENT_SUCCESS';
export const FETCH_COMMENT_FAIL = 'FETCH_COMMENT_FAIL';
export const fetchComment = id => {
  return {
    type: FETCH_COMMENT,
    payload: {
      request: {
        url: `/comments/${id}`
      }
    }
  };
};

// POST /comments
// Add a comment to a post.
export const ADD_COMMENT = 'ADD_COMMENT';
export const ADD_COMMENT_SUCCESS = 'ADD_COMMENT_SUCCESS';
export const ADD_COMMENT_FAIL = 'ADD_COMMENT_FAIL';
export const addComment = (parentId, values) => {
  const { author, body } = values;
  const data = {
    id: uuid(),
    timestamp: Date.now(),
    parentId,
    author,
    body
  };

  return {
    type: ADD_COMMENT,
    payload: {
      request: {
        method: 'POST',
        url: '/comments',
        data
      }
    }
  };
};

// PUT /comments/:id
// Edit the details of an existing comment.
export const EDIT_COMMENT = 'EDIT_COMMENT';
export const EDIT_COMMENT_SUCCESS = 'EDIT_COMMENT_SUCCESS';
export const EDIT_COMMENT_FAIL = 'EDIT_COMMENT_FAIL';
export const editComment = (id, data) => {
  return {
    type: EDIT_COMMENT,
    payload: {
      request: {
        method: 'PUT',
        url: `/comments/${id}`,
        data
      }
    }
  };
};

// DELETE /comments/:id
// Sets a comment's deleted flag to true.
export const REMOVE_COMMENT = 'REMOVE_COMMENT';
export const REMOVE_COMMENT_SUCCESS = 'REMOVE_COMMENT_SUCCESS';
export const REMOVE_COMMENT_FAIL = 'REMOVE_COMMENT_FAIL';
export const removeComment = id => {
  return {
    type: REMOVE_COMMENT,
    payload: {
      request: {
        method: 'DELETE',
        url: `/comments/${id}`
      }
    }
  };
};

// POST /comments/:id
// Used for voting on a comment.
export const VOTE_COMMENT = 'VOTE_COMMENT';
export const VOTE_COMMENT_SUCCESS = 'VOTE_COMMENT_SUCCESS';
export const VOTE_COMMENT_FAIL = 'VOTE_COMMENT_FAIL';
const vote = (id, type) => {
  return {
    type: VOTE_COMMENT,
    payload: {
      request: {
        method: 'POST',
        url: `/comments/${id}`,
        data: {
          option: type
        }
      }
    }
  };
};

export const voteCommentUp = id => {
  return vote(id, 'upVote');
};

export const voteCommentDown = id => {
  return vote(id, 'downVote');
};
