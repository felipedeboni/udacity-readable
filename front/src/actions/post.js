import uuid from 'utils/uuid';

// GET /posts
// Get all of the posts. Useful for the main page when no category is selected.
export const FETCH_ALL_POSTS = 'FETCH_ALL_POSTS';
export const FETCH_ALL_POSTS_SUCCESS = 'FETCH_ALL_POSTS_SUCCESS';
export const FETCH_ALL_POSTS_FAIL = 'FETCH_ALL_POSTS_FAIL';
export const fetchAllPosts = () => {
  return {
    type: FETCH_ALL_POSTS,
    payload: {
      request: {
        url: `/posts`
      }
    }
  };
};

// GET /:category/posts
// Get all of the posts for a particular category.
export const FETCH_POSTS_BY_CATEGORY = 'FETCH_POSTS_BY_CATEGORY';
export const FETCH_POSTS_BY_CATEGORY_SUCCESS =
  'FETCH_POSTS_BY_CATEGORY_SUCCESS';
export const FETCH_POSTS_BY_CATEGORY_FAIL = 'FETCH_POSTS_BY_CATEGORY_FAIL';
export const fetchPostsByCategory = categoryId => {
  return {
    type: FETCH_POSTS_BY_CATEGORY,
    payload: {
      request: {
        url: `/${categoryId}/posts`
      }
    }
  };
};

// GET /posts/:id
// Get the details of a single post.
export const FETCH_POST = 'FETCH_POST';
export const FETCH_POST_SUCCESS = 'FETCH_POST_SUCCESS';
export const FETCH_POST_FAIL = 'FETCH_POST_FAIL';
export const fetchPost = id => {
  return {
    type: FETCH_POST,
    payload: {
      request: {
        url: `/posts/${id}`
      }
    }
  };
};

// POST /posts
// Add a new post.
export const ADD_POST = 'ADD_POST';
export const ADD_POST_SUCCESS = 'ADD_POST_SUCCESS';
export const ADD_POST_FAIL = 'ADD_POST_FAIL';
export const addPost = ({ category, author, title, body }) => {
  const data = {
    id: uuid(),
    timestamp: Date.now(),
    category,
    author,
    title,
    body
  };

  return {
    type: ADD_POST,
    payload: {
      request: {
        method: 'POST',
        url: '/posts',
        data
      }
    }
  };
};

// PUT /posts/:id
// Edit the details of an existing post.
export const EDIT_POST = 'EDIT_POST';
export const EDIT_POST_SUCCESS = 'EDIT_POST_SUCCESS';
export const EDIT_POST_FAIL = 'EDIT_POST_FAIL';
export const editPost = (id, data) => {
  return {
    type: EDIT_POST,
    payload: {
      request: {
        method: 'PUT',
        url: `/posts/${id}`,
        data
      }
    }
  };
};

// DELETE /posts/:id
// Sets the deleted flag for a post to 'true'.
// Sets the parentDeleted flag for all child comments to 'true'.
export const REMOVE_POST = 'REMOVE_POST';
export const REMOVE_POST_SUCCESS = 'REMOVE_POST_SUCCESS';
export const REMOVE_POST_FAIL = 'REMOVE_POST_FAIL';
export const removePost = id => {
  return {
    type: REMOVE_POST,
    payload: {
      id: id,
      request: {
        method: 'DELETE',
        url: `/posts/${id}`
      }
    }
  };
};

// POST /posts/:id
// Used for voting on a post.
export const VOTE_POST = 'VOTE_POST';
export const VOTE_POST_SUCCESS = 'VOTE_POST_SUCCESS';
export const VOTE_POST_FAIL = 'VOTE_POST_FAIL';
const vote = (id, type) => {
  return {
    type: VOTE_POST,
    payload: {
      request: {
        method: 'POST',
        url: `/posts/${id}`,
        data: {
          option: type
        }
      }
    }
  };
};

export const votePostUp = id => {
  return vote(id, 'upVote');
};

export const votePostDown = id => {
  return vote(id, 'downVote');
};
