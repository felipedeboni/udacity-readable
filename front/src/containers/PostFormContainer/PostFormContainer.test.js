import React from 'react';
import { shallow } from 'enzyme';
import * as PostActions from 'actions/post';
import ConnectedPostFormContainer, {
  PostFormContainer
} from './PostFormContainer';

describe('PostFormContainer', () => {
  const { jestPromise } = global;

  const render = () => {
    const [post] = global.mock.posts;
    const state = {
      posts: global.mock.postsObject
    };

    const { component, store } = global.shallowWithStore(
      <ConnectedPostFormContainer
        categories={global.mock.categories}
        id={post.id}
      />,
      state
    );

    return {
      component,
      store,
      post
    };
  };

  it('renders without crashing', () => {
    const { component } = render();

    expect(component.dive()).toMatchSnapshot();
  });

  it('fetches the post', () => {
    const [post] = global.mock.posts;
    const { component, store } = global.shallowWithStore(
      <ConnectedPostFormContainer
        categories={global.mock.categories}
        id={post.id}
        match={{ params: { category: post.category } }}
        history={{ push: jest.fn(), replace: jest.fn() }}
      />,
      { posts: {} }
    );

    expect(component.dive()).toMatchSnapshot();
    expect(
      store.getActions().find(_ => _.type === PostActions.FETCH_POST)
    ).toBeDefined();
  });

  it('dispatch addPost', () => {
    const { component, store } = render();
    const form = component;

    form.props().add({});
    expect(
      store.getActions().find(_ => _.type === PostActions.ADD_POST)
    ).toBeDefined();
  });

  it('dispatch editPost', () => {
    const { component, store } = render();
    const form = component;

    form.props().edit({});
    expect(
      store.getActions().find(_ => _.type === PostActions.EDIT_POST)
    ).toBeDefined();
  });

  it('calls add on submit', () => {
    const addFn = jest
      .fn()
      .mockImplementation(() => new Promise(() => {}, () => {}));
    const component = shallow(
      <PostFormContainer categories={global.mock.categories} add={addFn} />
    );

    component.instance().onSubmit({});
    expect(addFn).toHaveBeenCalled();
  });

  it('calls edit on submit', () => {
    const [post] = global.mock.posts;

    const editFn = jest
      .fn()
      .mockImplementation(() => new Promise(() => {}, () => {}));
    const component = shallow(
      <PostFormContainer
        categories={global.mock.categories}
        id={post.id}
        post={post}
        edit={editFn}
      />
    );

    component.instance().onSubmit({});
    expect(editFn).toHaveBeenCalled();
  });

  it('should redirect after saving', () => {
    const addFn = jestPromise();
    const history = {
      push: jest.fn()
    };

    const component = shallow(
      <PostFormContainer
        categores={global.mock.categories}
        add={addFn}
        history={history}
      />
    );
    component.instance().afterSave({
      payload: {
        data: {
          id: 1,
          category: 'react'
        }
      }
    });
    component.instance().onSubmit({});
    expect(history.push).toHaveBeenCalledWith(`/react/1`);
  });
});
