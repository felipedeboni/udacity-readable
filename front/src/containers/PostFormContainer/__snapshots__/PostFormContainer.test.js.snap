// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostFormContainer fetches the post 1`] = `
<div>
  <div>
    Loading
  </div>
</div>
`;

exports[`PostFormContainer renders without crashing 1`] = `
<div>
  <C
    categories={
      Array [
        Object {
          "name": "react",
          "path": "react",
        },
        Object {
          "name": "redux",
          "path": "redux",
        },
        Object {
          "name": "udacity",
          "path": "udacity",
        },
      ]
    }
    id="8xf0y6ziyjabvozdd253nd"
    onSubmitHandler={[Function]}
    post={
      Object {
        "author": "thingtwo",
        "body": "Everyone says so after all.",
        "category": "react",
        "commentCount": 2,
        "deleted": false,
        "id": "8xf0y6ziyjabvozdd253nd",
        "timestamp": 1467166872634,
        "title": "Udacity is the best place to learn React",
        "voteScore": 6,
      }
    }
  />
</div>
`;
